File: server/src/modules/fractionalIndex.ts
const alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
const maxDigit = alphabet.length - 1
const maxLength = 128

const digitMap = new Map([...alphabet].map((char, index) => [char, index]))

export const FractionalIndex = {
  between,
  before,
  after,
  sequence,
  isValid,
}

function between(left?: string | null, right?: string | null): string {
  if (left != null && right != null && left >= right) {
    throw new Error("left fractional index must be lower than right")
  }

  let prefix = ""
  let position = 0

  while (true) {
    const leftDigit = digitAt(left, position, 0)
    const rightDigit = digitAt(right, position, maxDigit)

    if (rightDigit - leftDigit > 1) {
      const digit = Math.floor((leftDigit + rightDigit) / 2)
      return prefix + alphabet[digit]
    }

    prefix += alphabet[leftDigit]
    position += 1
  }
}

function before(first?: string | null): string {
  return between(null, first)
}

function after(last?: string | null): string {
  return between(last, null)
}

function sequence(count: number): string[] {
  const result: string[] = []
  let previous: string | null = null

  for (let index = 0; index < count; index += 1) {
    previous = after(previous)
    result.push(previous)
  }

  return result
}

function isValid(value: string): boolean {
  if (value.length === 0 || value.length > maxLength) {
    return false
  }

  for (const char of value) {
    if (!digitMap.has(char)) {
      return false
    }
  }

  return true
}

function digitAt(value: string | null | undefined, position: number, fallback: number): number {
  if (value == null || position >= value.length) {
    return fallback
  }

  const char = value[position]
  if (char == null) {
    return fallback
  }

  const digit = digitMap.get(char)
  if (digit == null) {
    throw new Error(`invalid fractional index character: ${char}`)
  }

  return digit
}

File: apple/InlineKit/Sources/InlineKit/Utils/FractionalIndex.swift
import Foundation

public enum FractionalIndex {
  private static let alphabet = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz".utf8)
  private static let maxDigit = alphabet.count - 1
  private static let digitMap = Dictionary(uniqueKeysWithValues: alphabet.enumerated().map { ($0.element, $0.offset) })

  public static func between(_ left: String?, _ right: String?) -> String {
    if let left, let right {
      precondition(left < right, "left fractional index must be lower than right")
    }

    let leftBytes = left.map { Array($0.utf8) }
    let rightBytes = right.map { Array($0.utf8) }
    var result: [UInt8] = []
    var position = 0

    while true {
      let leftDigit = digit(in: leftBytes, at: position, fallback: 0)
      let rightDigit = digit(in: rightBytes, at: position, fallback: maxDigit)

      if rightDigit - leftDigit > 1 {
        result.append(alphabet[(leftDigit + rightDigit) / 2])
        return String(decoding: result, as: UTF8.self)
      }

      result.append(alphabet[leftDigit])
      position += 1
    }
  }

  public static func before(_ first: String?) -> String {
    between(nil, first)
  }

  public static func after(_ last: String?) -> String {
    between(last, nil)
  }

  public static func sequence(count: Int) -> [String] {
    guard count > 0 else { return [] }

    var result: [String] = []
    var previous: String?

    for _ in 0..<count {
      let next = after(previous)
      result.append(next)
      previous = next
    }

    return result
  }

  private static func digit(in bytes: [UInt8]?, at position: Int, fallback: Int) -> Int {
    guard let bytes, position < bytes.count else {
      return fallback
    }

    guard let digit = digitMap[bytes[position]] else {
      preconditionFailure("invalid fractional index character")
    }

    return digit
  }
}

